### PR TITLE
Implement reflection for all components

### PR DESCRIPTION
**Pull request by RiscadoA**
# Description

Implements reflection for all components and for their fields.
Adds some utilities to make this easier.

Read this commit by commit, I split it nicely :relieved: 

## Checklist

- [x] Self-review changes.
- [x] Evaluate impact on the documentation.
- [x] Ensure test coverage.
- [x] ~~Write new samples.~~

### CODE REVIEW

File: core/include/cubos/core/ecs/component/reflection.hpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    public:
+        /// @brief Constructs.
+        /// @param name Component type name, including namespace.
+        ComponentTypeBuilder(std::string name)
+            : mType(reflection::Type::create(std::move(name)))
+        {
+            mType.with(reflection::ConstructibleTrait::typed<T>()
+                           .withDefaultConstructor()
+                           .withCopyConstructor()
+                           .withMoveConstructor()
```
Comment (luishfonseca) [inline core/include/cubos/core/ecs/component/reflection.hpp:32]: Adding these 3 constructors is so frequent we should have a short hand for them, maybe `.withBasicConstructors()`?

File: core/include/cubos/core/ecs/component/reflection.hpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            mType.with(std::move(mFields));
+            return mType;
+        }
+
+    private:
+        reflection::Type& mType;
+        reflection::FieldsTrait mFields;
+    };
+} // namespace cubos::core::ecs
```
Comment (luishfonseca) [inline core/include/cubos/core/ecs/component/reflection.hpp:60]: Missing trailing new line

File: core/include/cubos/core/reflection/traits/fields.hpp [hunk 3]
```diff
@@ -54,19 +54,30 @@ namespace cubos::core::reflection
         /// @param type Field type.
         /// @param name Field name.
         /// @param addressOf Getter for the address of the field on a given instance.
+        void addField(const Type& type, std::string name, AddressOf* addressOf);
+
+        /// @copydoc addField(const Type&, std::string, AddressOf*)
         /// @return Trait.
         FieldsTrait&& withField(const Type& type, std::string name, AddressOf* addressOf) &&;
```
Comment (luishfonseca) [inline core/include/cubos/core/reflection/traits/fields.hpp:61]: Why is this needed? The templated `withField` calls `addField`

**Review by luishfonseca — CHANGES_REQUESTED** [review]

File: core/include/cubos/core/reflection/traits/fields.hpp (hunk 3 above)
Comment (RiscadoA) [reply core/include/cubos/core/reflection/traits/fields.hpp:61]: We use in the `ComponentTypeBuilder` to avoid having to do this:
```cpp
mFieldsTrait = std::move(mFieldsTrait).withField(...);
```
This would mean having to implement the assignment operator for `FieldsTrait`. This way we can just do:
```cpp
mFieldsTrait.addField(...);
```

Comment (luishfonseca) [reply core/include/cubos/core/reflection/traits/fields.hpp:61]: The review was about the 3 argument `withField`, as I tought it was dead code as both templated `addField` and `withField` used the 3 argument `addField`.
This isn't the case, glm reflection uses 3 argument `withField` so it makes sense to keep it.

**Review by luishfonseca — APPROVED** [review]
